### ISSUE TITLE
Autodoc: links to `AnyReader` and `AnyWriter` do not resolve correctly

### ISSUE THREAD

**Issue by ianprime0509** [ziglang/zig#19528]
### Zig Version

0.12.0-dev.3522+b88ae8dbd

### Steps to Reproduce and Observed Behavior

Run `zig std` and navigate to `#std.io`. Then, click on `AnyReader` or `AnyWriter` in the "types" section at the top. The links end up leading to `std.io.GenericReader` and `std.io.GenericWriter`, respectively.

### Expected Behavior

The links should lead to the appropriate types.

This peculiar situation is caused because the links are constructed as `#std.io.Reader` and `#std.io.Writer` following this logic, which is relying on the file paths to the resolved decls: https://github.com/ziglang/zig/blob/b88ae8dbd84886d3b9b26509034720f755a0e28a/lib/docs/wasm/Decl.zig#L140-L190 (the underlying decls here being the file-level structs `std/io/Reader.zig` and `std/io/Writer.zig`). But when resolving a path to a decl, priority is given to the component-wise interpetation, under which `std.io.Reader` and `std.io.Writer` are resolved to the `Reader` and `Writer` decls in `std.io`, which are actually deprecated aliases for `GenericReader` and `GenericWriter`: https://github.com/ziglang/zig/blob/b88ae8dbd84886d3b9b26509034720f755a0e28a/lib/docs/wasm/main.zig#L839-L859

**Comment by travisstaloch** [comment]
Some links for context:

[std.io autodocs](https://ziglang.org/documentation/master/std/#std.io)

[AnyReader src](https://github.com/ziglang/zig/blob/b88ae8dbd84886d3b9b26509034720f755a0e28a/lib/std/io.zig#L368)

Thanks @ianprime0509 for reporting this and for helping me to understand that this was actually a bug on discord.  I was confused by the aliasing.  But my first impression that AnyReader links landing on GenericReader ended up being correct.

**Comment by boomlinde** [comment]
autodoc seems to confuse file names with exports in some way. Without addressing or investigating the root cause, this specific problem seems like it can be solved by renaming `lib/std/io/Reader.zig` to `lib/std/io/AnyReader.zig`, `lib/std/io/Writer.zig` to `lib/std/io/AnyWriter.zig` and updating the imports of these files in `lib/std/io.zig` and CMakeLists.txt

Should I push a PR for such a workaround? I don't have a working build environment setup for zig so I have actually only tested this by updating the standard library in one of the binary distributions and verified with `zig std`.

**Comment by andrewrk** [comment]
> Should I push a PR for such a workaround?

no, only addressing or investigating the root cause efforts will be entertained

**Comment by achan1989** [comment]
Links in JS are effectively the fully qualified name of the thing you want to view. When you navigate to a link, the JS asks the WASM to resolve the FQN to a decl, then the JS renders that decl.

The problem is that an FQN does not uniquely identify a decl. `std.io.Reader` is the FQN given to `std.io.AnyReader`, because it's implemented in `std.io.Reader.zig`; but `std.io.Reader` is also the FQN given to an alias of `std.io.GenericReader`, because `pub const Writer = GenericWriter` in `std.io.zig`.

A few ways we could address this:
1. Forbid situations that cause non-unique FQNs to be generated under the current scheme. Andrew has indicated this is a no-go.
1. Change the format of FQNs, and/or how they're generated, such that they're actually unique. I haven't thought about how that could work, and I don't know if there are other users of FQNs which would be affected by such a change.
1. Change autodoc so that links use the guaranteed-unique decl index, not the FQN. So `https://ziglang.org/documentation/master/std/#std.io.AnyReader` would become like `https://ziglang.org/documentation/master/std/#1234`. It's a quick and easy fix. But the URL becomes opaque. And the decl indexes aren't stable between builds, so a given URL for the master branch docs could display completely different items over time. Seems like a non-starter.
1. Change autodoc so that links use some kind of augmented FQN. Maybe something like `https://ziglang.org/documentation/master/std/#typefn:std.io.Reader` would resolve to `std.io.GenericReader`, and `#container:std.io.Reader` would resolve to the `AnyReader` content for `std.io.Reader.zig`? I can't think of a scheme that isn't really unwieldy while still guaranteeing unique results under degenerate conditions.
